### PR TITLE
Feature: api backend on php8

### PR DESCRIPTION
**Pull request by DerekCrannaford**
## Checklist
- [ Y] I'm using the correct PHP Version (8.1).
- [ Y] I've added comments to any new methods I've created or where else relevant.
- [ Y] I've replaced magic method usage on DbService classes with the getInstance() static method.
- [ Y] I've written any documentation for new features or where else relevant in the docs [repo](https://github.com/2pisoftware/cmfive-docs).

## Description
Settle Token auth (API) support into PHP8.1 core

## Changelog
Formalise stable tokens/auth 'hooks' subsystem for actions
Remove all test & demo code from tokens/auth module(s)

### CODE REVIEW

File: system/modules/tokens/models/TokensService.php [hunk 1]
```diff
@@ -0,0 +1,67 @@
+<?php
+
+/**@author Derek Crannaford */
+
+class TokensService extends DbService
+{
+
+    public function getTokenFromAuthorisationHeader($bearer)
```
Comment (adam-buckley) [inline system/modules/tokens/models/TokensService.php:8]: Please add return types to all of the service functions that dont override a DbService function

File: system/modules/tokens/models/TokensService.php [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    }
+
+    function getBase64URL($plainText)
+    {
+        $base64 = base64_encode($plainText);
+        $base64 = trim($base64, "=");
+        $base64url = strtr($base64, '+/', '-_');
+        return ($base64url);
+    }
+
```
Comment (adam-buckley) [inline system/modules/tokens/models/TokensService.php:65]: Please clean up whitespace

File: system/modules/tokens/models/TokensPolicy.php [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+
+        // So, from here, let's hit another hook, for validators/app_service models to populate roles from profile:
+
+        // broadcast this policy
+        // listener will self identify as _validator and _app_id
+        // listener will interpret profile identifier (eg as user_id, group_policy, code-baked app_actions etc)
+        $hook_results = $this->w->callHook("tokens", "get_roles_from_policy", $this);
+
+        if (empty($hook_results)) {
+            $this->Log->error("No Roles for policy:" . $this->_role_profile);
```
Comment (adam-buckley) [inline system/modules/tokens/models/TokensPolicy.php:43]: Please remove magic service method usage

File: system/modules/tokens/models/TokensPolicy.php [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+     * in the system.
+     *
+     * This will execute all the functions associated to the bearer's roles
+     * until one function returns true.
+     *
+     * @param Web $w
+     * @param string $path
+     * @return true if one role function returned true
+     */
+    public function tokensAllowed($path)
```
Comment (adam-buckley) [inline system/modules/tokens/models/TokensPolicy.php:68]: Please add function return types

File: system/modules/tokens/models/ApiOutputService.php [hunk 5]
```diff
@@ -0,0 +1,94 @@
+<?php
+
+/**@author Derek Crannaford */
+
+class ApiOutputService extends DbService
+{
+    public function useNoTemplate($w)
+    {
+        $w->setLayout(null);
```
Comment (adam-buckley) [inline system/modules/tokens/models/ApiOutputService.php:9]: A one-line function is semi-redundant, please remove this

File: system/modules/tokens/models/ApiOutputService.php [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+    public function apiReturnCmfiveStyledHtml($w, $response) {
```
Comment (adam-buckley) [inline system/modules/tokens/models/ApiOutputService.php:12]: Please add function return types

File: system/modules/tokens/models/ApiOutputService.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+            $w->setLayout("layout-modal");
+            $w->enqueueScript(array("name" => "main.js", "uri" => "/system/templates/js/main.js", "weight" => 995));
```
Comment (adam-buckley) [inline system/modules/tokens/models/ApiOutputService.php:15]: I wonder if we can "future proof" this a bit with an enum or something of loadable js and css, that way we can switch to the new designs? Just an idea

File: system/modules/tokens/models/ApiOutputService.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+            $w->enqueueStyle(array("name" => "foundation.css", "uri" => "/system/templates/js/foundation-5.5.0/css/foundation.css", "weight" => 1005));
```
Comment (adam-buckley) [inline system/modules/tokens/models/ApiOutputService.php:16]: Please remove long array syntax

File: system/modules/tokens/models/ApiOutputService.php [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+        // mark header for return content type JSON
+        if (substr($response['status'],0,1) == "2") {
+            header('Content-Type: application/json');
+        } else {
+            LogService::getInstance($this->w)->info("API request rejected: ".$response['referer']);
+            // Don't need, already have set response code!
+            // header($_SERVER["SERVER_PROTOCOL"] . " " . $response['status'] . " " . $response['payload'][0]);
+        }
+
+        echo json_encode(['response' => $response]);
```
Comment (adam-buckley) [inline system/modules/tokens/models/ApiOutputService.php:38]: We have a JsonResponse class that should be able to do most of what this function is doing

File: system/config.php [hunk 10]
```diff
@@ -94,3 +94,5 @@
         'secret' => '',
     ],
 ]);
+
+Config::set('system.use_api', true);
```
Comment (adam-buckley) [inline system/config.php:98]: Should this be defaulted to false?

File: system/html.php [hunk 11]
```diff
@@ -365,8 +365,8 @@ public static function form($data, $action = null, $method = "POST", $submitTitl
             $readonly = "";
 
             // handle disabled fields
-            if (substr($name, 0, 1) == '-') {
-                $name = substr($name, 1);
+            if (substr(($name??""), 0, 1) == '-') {
```
Comment (adam-buckley) [inline system/html.php:368]: Please add spacing around ```??```

**Review by adam-buckley — DISMISSED** [review]

File: system/modules/tokens/models/ApiOutputService.php (hunk 6 above)
Comment (DerekCrannaford) [reply system/modules/tokens/models/ApiOutputService.php:12]: These functions all result in spooled output, then termination of PHP.

File: system/config.php (hunk 10 above)
Comment (DerekCrannaford) [reply system/config.php:98]: "true" seems OK as the default, no change to auth access will result unless additional modules also implement hook reactions.

"false" will also inhibit authModule from ever firing: ->callHook "get_auth_token_validation"
Which is not much processing overhead in the case of no hooks existing?

File: system/html.php (hunk 11 above)
Comment (DerekCrannaford) [reply system/html.php:368]: OK - they're a bit left of field :)

File: system/modules/tokens/models/ApiOutputService.php (hunk 9 above)
Comment (DerekCrannaford) [reply system/modules/tokens/models/ApiOutputService.php:38]: Fair point - in the category "I would have used that if I'd thought of it."
Meanwhile, stuck with this one:
 - APP's in the wild are already expecting "PAYLOAD" not "DATA"
 - It logs an error with IP, if the API is maybe being abused
 - Status codes can be elaborated on last digit(s) if desired
 
 If writing a new API/APP, results are returned by any means of ACTIONS = JsonResponse class, cmfive .tpl's etc are all available choices.

File: system/modules/tokens/models/ApiOutputService.php (hunk 7 above)
Comment (DerekCrannaford) [reply system/modules/tokens/models/ApiOutputService.php:15]: As discussed elsewhere, if core style/tpl's got their own ComponentRegister / enqueue requirements programatically, we could reflect that here. Currently it means copy-paste long hand from the .tpl with favoured style, over to here, to be applied to modal-as-blank. Other option would be to declare several "blank" tpl's aside from modal - each being basically just the Register/Enqueue, with no additional page layout.

File: system/modules/auth/models/AuthService.php [hunk 12]
```diff
@@ -328,11 +327,52 @@ public function allowed($path, $url = null)
             return false;
         }
 
+        // Whitelisted action, or white-bread login session
         if ((function_exists("anonymous_allowed") && anonymous_allowed($this->w, $path)) || ($this->user() && $this->user()->allowed($path))) {
             self::$_cache[$key] = $url ? $url : true;
             return self::$_cache[$key];
         }
 
+        // API token handling - WIP 202203
```
Comment (Dane-2pi) [inline system/modules/auth/models/AuthService.php:336]: This probably isn't WIP, nor that date any more.

File: system/modules/auth/models/AuthService.php [hunk 13]
```diff
@@ -346,6 +386,10 @@ public function allowed($path, $url = null)
                     $this->forceLogin($user->id);
                     if ($user->allowed($path)) {
                         self::$_cache[$key] = $url ? $url : true;
+                        // hmmm, so we have forced login,
```
Comment (Dane-2pi) [inline system/modules/auth/models/AuthService.php:389]: clean up this commenting too

File: system/web.php [hunk 14]
```diff
@@ -157,7 +157,7 @@ private function getUrlSchema()
 
     private function getHostname()
     {
-        return $this->getRequestHeader('HTTP_HOST');
+        return $this->getRequestHeader("HTTP_X_FORWARDED_HOST", $this->getRequestHeader('HTTP_HOST'));
```
Comment (Dane-2pi) [inline system/web.php:160]: does this work universally ? 

It's in place for use on codespaces. And if it works everywhere, then it'd be good to have it in web by default, but I was under the impression that it didn't, and should only be used for dev on codespaces.

Comment (Dane-2pi) [inline system/web.php:160]: does this work universally ? 

It's in place for use on codespaces. And if it works everywhere, then it'd be good to have it in web by default, but I was under the impression that it didn't, and should only be used for dev on codespaces.

Comment (Dane-2pi) [reply system/web.php:160]: Fair, I guess we can roll it in, and I'll stop cherypicking that change for codespaces then.

**Review by Dane-2pi — APPROVED** [review]
Looks clean, if it's been tested I'm happy to approve it.

**Pull request merged**